protocol/swap/manager.go

```diff
@@ -1,168 +1,156 @@
-}
-
-// Status returns the swap's status.
-func (i *Info) Status() Status {
-	return i.status
-}
-
-// StatusCh returns the swap's status update channel.
-func (i *Info) StatusCh() <-chan types.Status {
-	return i.statusCh
-}
-
-// SetStatus ...
-func (i *Info) SetStatus(s Status) {
-	i.status = s
-}
+var errNoSwapWithID = errors.New("unable to find swap with given ID")
 
 // Manager tracks current and past swaps.
 type Manager interface {
 	AddSwap(info *Info) error
-	GetPastIDs() []types.Hash
-	GetPastSwap(types.Hash) *Info
-	GetOngoingSwap(types.Hash) *Info
-	CompleteOngoingSwap(types.Hash)
+	GetPastIDs() ([]types.Hash, error)
+	GetPastSwap(types.Hash) (*Info, error)
+	GetOngoingSwap(types.Hash) (*Info, error)
+	CompleteOngoingSwap(types.Hash) error
 }
 
 type manager struct {
+	db Database
 	sync.RWMutex
 	ongoing map[types.Hash]*Info
 	past    map[types.Hash]*Info
 }
 
-// NewManager ...
-func NewManager() Manager {
+// NewManager returns a new Manager that uses the given database.
+// It loads all ongoing swaps into memory on construction.
+// Completed swaps are not loaded into memory.
+func NewManager(db Database) (*manager, error) {
+	ongoing := make(map[types.Hash]*Info)
+
+	stored, err := db.GetAllSwaps()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, s := range stored {
+		if !s.Status.IsOngoing() {
+			continue
+		}
+
+		ongoing[s.ID] = s
+	}
+
 	return &manager{
-		ongoing: make(map[types.Hash]*Info),
+		db:      db,
+		ongoing: ongoing,
 		past:    make(map[types.Hash]*Info),
-	}
+	}, nil
 }
 
 // AddSwap adds the given swap *Info to the Manager.
 func (m *manager) AddSwap(info *Info) error {
 	m.Lock()
 	defer m.Unlock()
 
-	switch info.status.IsOngoing() {
+	switch info.Status.IsOngoing() {
 	case true:
-		m.ongoing[info.id] = info
+		m.ongoing[info.ID] = info
 	default:
-		m.past[info.id] = info
+		m.past[info.ID] = info
 	}
 
-	return nil
+	return m.db.PutSwap(info)
 }
 
 // GetPastIDs returns all past swap IDs.
-func (m *manager) GetPastIDs() []types.Hash {
+func (m *manager) GetPastIDs() ([]types.Hash, error) {
 	m.RLock()
 	defer m.RUnlock()
-	ids := make([]types.Hash, len(m.past))
-	i := 0
+	ids := make(map[types.Hash]struct{})
 	for id := range m.past {
-		ids[i] = id
+		ids[id] = struct{}{}
+	}
+
+	// TODO: do we want to cache all past swaps since we're already fetching them?
+	stored, err := m.db.GetAllSwaps()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, s := range stored {
+		if s.Status.IsOngoing() {
+			continue
+		}
+
+		ids[s.ID] = struct{}{}
+	}
+
+	idArr := make([]types.Hash, len(ids))
+	i := 0
+	for id := range ids {
+		idArr[i] = id
 		i++
 	}
-	return ids
+
+	return idArr, nil
 }
 
 // GetPastSwap returns a swap's *Info given its ID.
-func (m *manager) GetPastSwap(id types.Hash) *Info {
+func (m *manager) GetPastSwap(id types.Hash) (*Info, error) {
 	m.RLock()
 	defer m.RUnlock()
-	return m.past[id]
+	s, has := m.past[id]
+	if has {
+		return s, nil
+	}
+
+	// TODO: do we want to cache this swap?
```

Comment:
I would say yes.  If you are going to cache any, a most recently used algorithm would be the way to go. If the database is fast, we may not need a cache at all. We are only serving a single user, so simplicity beats scalability.